File: Algorithms/143.ReorderList.cpp
class Solution {
public:
    void reorderList(ListNode *head) {
        ListNode* p = split(head);
        p = reverse(p);
        head = merge(head, p);
    }
    ListNode* split(ListNode *head) {
        if (head == NULL) 
            return head;
        ListNode *slow = head;
        ListNode *fast = head->next;
        while (fast != NULL && fast->next != NULL) {
            slow = slow->next;
            fast = fast->next->next;
        }
        ListNode *p = slow->next;
        slow->next = NULL;
        return p;
    }
    ListNode* reverse(ListNode *head) { 
        ListNode *headptr = new ListNode(0);
        while (head != NULL) {
            ListNode *tmp = head->next;
            head->next = headptr->next;
            headptr->next = head;
            head = tmp;
        }
        return headptr->next;
    }
    ListNode* merge(ListNode *p, ListNode *q) {
        ListNode *head = new ListNode(0);
        ListNode *tmp = head;
        while (p != NULL && q != NULL) {
            tmp->next = p;
            p = p->next;
            tmp = tmp->next;
            tmp->next = q;
            q = q->next;
            tmp = tmp->next;
        }
        if (p != NULL) {
            tmp->next = p;
        }
        return head->next;
    }
};

File: 143.ReorderList.py
class Solution:
    # @param head, a ListNode
    # @return a ListNode
    def reorderList(self, head):
        p = self.split(head)
        p = self.reverse(p)
        head = self.merge(head, p)
        return head

    def split(self, head):
        if not head:
            return head
        slow = head
        fast = head.next
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        p = slow.next
        slow.next = None
        return p

    def reverse(self, head):
        headptr = ListNode(0)
        while head:
            tmp = head.next
            head.next = headptr.next
            headptr.next = head
            head = tmp
        return headptr.next

    def merge(self, p, q):
        head = ListNode(0)
        tmp = head
        while p and q:
            tmp.next = p
            p = p.next
            tmp = tmp.next
            tmp.next = q
            q = q.next
            tmp = tmp.next
        if p:
            tmp.next = p
        return head.next
